Benchmark updates

agibsonccc opened the pull request:
## What changes were proposed in this pull request?

Adds some updates for nd4j benchmarking to add memory pressure test and updated documentation for using the benchmark with jemalloc

Adds opaquendarray caching and deletion on close()
Add new reduce smoke test
## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.konduit.ai/multi-project/how-to-guides/contribute/eclipse-contributors) page for details
- [ X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X ] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

agibsonccc committed:
add new smoke tests
fix scalar index when using the indexing api
fix usage of index2coords by migrating shape information
to shape::shapeOf
fix usage of coords2index by migrating accidental shapeOf invocations to strides

agibsonccc committed:
Fix reduce3 buffer() calls in op execution

agibsonccc committed:
Add opaquendarray cachcing/deletion  on close
Add new memory pressure test
Update nd4j benchmarking with jemalloc docs
Add licenses
Remove more old aurora code
Add basic reduce smoke tests

agibsonccc committed:
Merge remote-tracking branch 'origin/master' into ag_benchmark_updates

# Conflicts:
#	libnd4j/include/legacy/cpu/NativeOps.cpp
#	libnd4j/include/ops/ops.h
#	nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ndarray/BaseNDArray.java
#	platform-tests/src/test/java/org/eclipse/deeplearning4j/nd4j/linalg/BroadcastingOpsSmokeTests.java

agibsonccc committed:
remove print statements

agibsonccc committed:
add more debugging documentaiton under troubleshooting/
rewrite benchmarks
Add new memory profiler
Fix deallocation crash
Add environment sourcing in the custom java executable to allow for more customization

treo commented on contrib/benchmarking_nd4j/pom.xml:
```diff
@@ -83,8 +88,8 @@
                 <version>3.8.1</version>
                 <configuration>
                     <compilerVersion>${javac.target}</compilerVersion>
-                    <source>${javac.target}</source>
-                    <target>${javac.target}</target>
```
This looks a bit off, you bothch change the parameter as part of the properties, and then change it here to a different fixed value. I guess that isn't quite intentional?

treo commented on contrib/benchmarking_nd4j/src/main/java/org/nd4j/memorypressure/MemoryPressureTest.java:
```diff
@@ -10,12 +10,15 @@ public class MemoryPressureTest {
-                    INDArray arrTwo = Nd4j.ones(200);
+
                     arrOne.addi(arrTwo);
+                    System.out.println("Completed " + atomicInteger.get() + " iterations");
```
Is this intentional that you create 100000 prints here?

agibsonccc replied:
Yes this one runs for a LONG time and the main use case isn't throughput here but GC.

treo commented on platform-tests/bin/java:
```diff
@@ -23,6 +23,7 @@
 set -exo pipefail
 #JAVA_CALL="java -DsourceCodeIndexerPath=/home/agibsonccc/Documents/GitHub/deeplearning4j -javaagent:/home/agibsonccc/.m2/repository/org/nd4j/nd4j-log-analyzer/1.0-SNAPSHOT/nd4j-log-analyzer-1.0-SNAPSHOT.jar "
 JAVA_CALL="java "
+source /home/agibsonccc/Documents/GitHub/deeplearning4j/platform-tests/bin/env.sh
```
It seems wrong to keep your path here.

agibsonccc replied:
Ah yeah this was left over from some troubleshooting. I wasn't sure if the variables were working or not so I was double checking paths..

agibsonccc committed:
add more debugging documentaiton under troubleshooting/
rewrite benchmarks
Add new memory profiler
Fix deallocation crash
Add environment sourcing in the custom java executable to allow for more customization
